### ISSUE TITLE
dbus doesn't build with clang on OpenBSD

### ISSUE THREAD

**Issue by blackgnezdo** [rblaze/haskell-dbus#43]
cabal v2-build for dbus fails with a puzzling error:
```
[ 2 of 14] Compiling DBus.Internal.Types ( lib/DBus/Internal/Types.hs, /home/greg/s/dbus-1.2.3/dist-newstyle/build/x86_64-openbsd/ghc-8.6.4/dbus-1.2.3/build/DBus/Internal/Types.o )
                                                                                             
lib/DBus/Internal/Types.hs:469:308: error:                                                   
    Not in scope: type constructor or class ‘HsType’                                         
    Perhaps you meant ‘Type’ (line 51)                                                       
    |                                         
469 | IS_ATOM(Bool,       AtomBool,       TypeBoolean)                                       
```
After some head scratching the error reduces to a difference in a pair of invocations on this pared down input file:
```
#define IS_ATOM(HsType, AtomCons, TypeCons) \
    instance IsAtom HsType where \
    { }; \
    instance IsValue HsType where \
    { typeOf' _ = TypeCons \
    }; \
    instance IsVariant HsType where \
    { }

IS_ATOM(Bool,       AtomBool,       TypeBoolean)
```

cpphs generates a working version:
```
$ cpphs y
#line 1 "y"

instance IsAtom Bool where     { };     instance IsValue Bool where     { typeOf' _ = TypeBoolean     };     instance IsVariant Bool where     { }
```

whereas clang `cc -E` generates this beauty:
```
$ cc --version ; cc -E  -x assembler-with-cpp /tmp/y
OpenBSD clang version 8.0.1 (tags/RELEASE_801/final) (based on LLVM 8.0.1)
Target: amd64-unknown-openbsd6.6
Thread model: posix
InstalledDir: /usr/bin
# 1 "/tmp/y"
# 1 "<built-in>" 1
# 1 "/tmp/y" 2

instance IsAtom Bool where { }; instance IsValue Bool where { typeOf' _ = TypeCons };     instance IsVariant HsType where     { }
```

Notice the bashfully preserved HsType in the `cc -E` output. It's due to the apostrophe in `typeOf'` which screws up the tokenization in the 40 year old catastrophe also known as CPP. 

I also tested with `gcc -E` 4.2.1 and 8.3.0 and they too generate broken code:
```
egcc --version; egcc -E  -x assembler-with-cpp /tmp/y
egcc (GCC) 8.3.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

# 1 "/tmp/y"
# 1 "<built-in>"
# 1 "<command-line>"
# 1 "/tmp/y"
# 9 "/tmp/y"
instance IsAtom Bool where { }; instance IsValue Bool where { typeOf' _ = TypeCons };     instance IsVariant HsType where     { }
```

**Comment by blackgnezdo** [comment]
FWIW, gcc -E can be made to digest the code as is by giving it `-traditional` which is probably why the problem went unnoticed for this long. Clang is less permissive even with `-traditional` and hence my problem. Both gcc and clang complain about the issue without `-traditional`:
```
set -x; cat y; egcc --version ; egcc -E  -traditional -undef - < /tmp/y;  egcc -E  -undef - < /tmp/y; cc --version ; cc -E  -undef - < /tmp/y
+ cat y
#define IS_ATOM(BAD) \
     BAD \
    { typeOf' _ = X }; \
     BAD \
    { }

IS_ATOM(Good)
+ egcc --version
egcc (GCC) 8.3.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

+ egcc -E -traditional -undef -
+ < /tmp/y 
# 1 "<stdin>"
# 1 "<built-in>"
# 1 "<command-line>"
# 1 "<stdin>"
Good     { typeOf' _ = X };      Good     { }
+ egcc -E -undef -
+ < /tmp/y 
# 1 "<stdin>"
# 1 "<built-in>"
# 1 "<command-line>"
# 1 "<stdin>"
<stdin>:3:13: warning: missing terminating ' character
Good { typeOf' _ = X };      BAD     { }
+ cc --version
OpenBSD clang version 8.0.1 (tags/RELEASE_801/final) (based on LLVM 8.0.1)
Target: amd64-unknown-openbsd6.6
Thread model: posix
InstalledDir: /usr/bin
+ cc -E -undef -
+ < /tmp/y 
# 1 "<stdin>"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "<stdin>" 2
<stdin>:3:13: warning: missing terminating ' character [-Winvalid-pp-token]
    { typeOf' _ = X }; \
            ^

Good { typeOf' _ = X };      BAD     { }
1 warning generated.

```

**Comment by blackgnezdo** [comment]
Could you release this fix to hackage at your convenience?

**Comment by rblaze** [comment]
Released `dbus-1.2.12`
